solver/llbsolver/solver.go

```diff
@@ -193,138 +197,189 @@ func (s *Solver) Solve(ctx context.Context, id string, sessionID string, req fro
+						return errors.Errorf("invalid reference: %T", r.Sys())
 					}
-					m[k] = workerRef.ImmutableRef
+					ctx = withDescHandlerCacheOpts(ctx, workerRef.ImmutableRef)
+
+					// Configure compression
+					compressionConfig := exp.Config().Compression
+
+					// all keys have same export chain so exporting others is not needed
+					_, err = r.CacheKeys()[0].Exporter.ExportTo(ctx, exp, solver.CacheExportOpt{
+						ResolveRemotes: workerRefResolver(cacheconfig.RefConfig{Compression: compressionConfig}, false, g),
+						Mode:           exp.CacheExportMode,
+						Session:        g,
+						CompressionOpt: &compressionConfig,
+					})
+					return err
+				}); err != nil {
+					return prepareDone(err)
 				}
-			}
-			inp.Refs = m
-		}
-
-		if _, ok := asInlineCache(exp.CacheExporter); ok {
-			if err := inBuilderContext(ctx, j, "preparing layers for inline cache", j.SessionID+"-cache-inline", func(ctx context.Context, _ session.Group) error {
-				if cached.Ref != nil {
-					dtic, err := inlineCache(ctx, exp.CacheExporter, cached.Ref, e.Config().Compression, session.NewGroup(sessionID))
-					if err != nil {
-						return err
-					}
-					if dtic != nil {
-						inp.AddMeta(exptypes.ExporterInlineCache, dtic)
-					}
+				prepareDone(nil)
+				resp, err := exp.Finalize(ctx)
+				select {
+				case respCh <- resp:
+				case <-ctx.Done():
 				}
-				for k, res := range cached.Refs {
-					dtic, err := inlineCache(ctx, exp.CacheExporter, res, e.Config().Compression, session.NewGroup(sessionID))
-					if err != nil {
-						return err
-					}
-					if dtic != nil {
-						inp.AddMeta(fmt.Sprintf("%s/%s", exptypes.ExporterInlineCache, k), dtic)
-					}
+				if err != nil {
+					return err
 				}
-				exp.CacheExporter = nil
 				return nil
-			}); err != nil {
-				return nil, err
+			})
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	close(respCh)
+	for resp := range respCh {
+		for k, v := range resp {
+			if cacheExporterResponse == nil {
+				cacheExporterResponse = make(map[string]string)
 			}
-		}
-		if err := inBuilderContext(ctx, j, e.Name(), j.SessionID+"-export", func(ctx context.Context, _ session.Group) error {
-			exporterResponse, err = e.Export(ctx, inp, j.SessionID)
-			return err
-		}); err != nil {
-			return nil, err
+			cacheExporterResponse[k] = v
 		}
 	}
+	return cacheExporterResponse, nil
+}
 
-	g := session.NewGroup(j.SessionID)
-	var cacheExporterResponse map[string]string
-	if e := exp.CacheExporter; e != nil {
-		if err := inBuilderContext(ctx, j, "exporting cache", j.SessionID+"-cache", func(ctx context.Context, _ session.Group) error {
-			prepareDone := progress.OneOff(ctx, "preparing build cache for export")
-			if err := res.EachRef(func(res solver.ResultProxy) error {
+func exportInlineCache(ctx context.Context, e exporter.ExporterInstance, inlineExporter *RemoteCacheExporter, j *solver.Job, res *frontend.Result) (exporterResponse map[string]string, err error) {
```

Comment:
The name might not be the most appropriate - I can rename to `exportWithInlineCacheExporter` - otherwise I believe it already contains the parts only relevant for the (non-cache) export.

Reply:
The name is better for sure - I think we should split up the function to only contain the inline cache export though and to do the non-cache export completely separately. There is some dependency there, but we can do the inline cache export by modifying a pointer to an `exporter.Source` object.